### PR TITLE
[#Ic-427]  update get message openapi

### PR DESCRIPTION
**Pull request by gquadrati**
This PR depends on #199 

#### List of Changes

- Update getMessage handler to return `readStatus` poperty when client got the right autorization group `ApiMessageReadAdvanced`
- Update unit and integration tests accordingly

#### Motivation and Context

New ADVANCED messages features

#### How Has This Been Tested?

Unit and integration tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: GetMessage/__tests__/handler.test.ts [hunk 1]
```diff
@@ -121,6 +124,18 @@ describe("GetMessageHandler", () => {
     userId: "u123" as NonEmptyString
   };
 
+  const aUserAuthenticationTrustedApplicationWithAdvancedFetures: IAzureApiAuthorization = {
+    groups: new Set([
+      UserGroup.ApiMessageRead,
+      UserGroup.ApiMessageList,
+      UserGroup.ApiMessageReadAdvanced,
+      UserGroup.ApiMessageReadAdvanced
```
Comment (AleDore) [inline GetMessage/__tests__/handler.test.ts:132]:
```suggestion

```

File: GetMessage/handler.ts [hunk 2]
```diff
@@ -117,6 +127,47 @@ type IGetMessageHandler = (
 const LegalMessagePattern = t.interface({ legal_data: LegalData });
 type LegalMessagePattern = t.TypeOf<typeof LegalMessagePattern>;
 
+/**
+ * Checks whether the client service can read advanced message info (read_status and payment_Status)
+ */
+export const canReadAdvancedMessageInfo = (
+  message:
+    | ExternalCreatedMessageWithoutContent
```
Comment (fabriziopapi) [inline GetMessage/handler.ts:135]: why a duplicated union type?

File: GetMessage/handler.ts [hunk 3]
```diff
@@ -117,6 +127,47 @@ type IGetMessageHandler = (
...
+  authGroups.has(UserGroup.ApiMessageReadAdvanced);
+
+/**
+ * Checks whether the client service can read message read status
+ *
+ * @param serviceId the subscription id of the service
+ * @returns false if user revoked the permission to access the read status, true otherwise
+ */
+export const canReadMessageReadStatus = (_serviceId: ServiceId): boolean =>
+  false;
```
Comment (fabriziopapi) [inline GetMessage/handler.ts:149]: maybe we could add a TODO comment here: something like "waiting for opt-out definition on profiles domain"

File: GetMessage/handler.ts [hunk 4]
```diff
@@ -252,19 +303,33 @@ export function GetMessageHandler(
...
+          maybeMessageStatus,
+          O.map(messageStatus => messageStatus.status),
+          // when the message has been received but a MessageStatus
+          // does not exist yet, the message is considered to be
+          // in the ACCEPTED state (not yet stored in the inbox)
+          O.getOrElse(() => MessageStatusValueEnum.ACCEPTED)
+        )
+      },
+      // Enrich message info with advanced properties if user is allowed to read them
+      messageWithoutAdvacedProperties =>
```
Comment (arcogabbo) [inline GetMessage/handler.ts:321]: typo in the variable name, propagated in lines 323 and 325

File: GetMessage/handler.ts [hunk 5]
```diff
@@ -117,6 +127,47 @@ type IGetMessageHandler = (
...
+  maybeMessageStatus: O.Option<RetrievedMessageStatus>
+): ReadStatus =>
+  pipe(
+    maybeMessageStatus,
+    O.map(messageStatus =>
+      b.fold(
+        () => ReadStatusEnum.UNREAD,
+        () => ReadStatusEnum.READ
+      )(messageStatus.isRead)
+    ),
```
Comment (fabriziopapi) [inline GetMessage/handler.ts:167]:
```suggestion
    O.map(messageStatus => messageStatus.isRead),
    O.map(
      b.fold(
        () => ReadStatusEnum.UNREAD,
        () => ReadStatusEnum.READ
      )
    ),
```
maybe we could flatten the inner fold

File: GetMessage/handler.ts (hunk 2 above)
Comment (gquadrati) [reply GetMessage/handler.ts:136]: Nice catch, thanks. Changed to `ExternalCreatedMessageWithContent` in [20fae5a](https://github.com/pagopa/io-functions-services/pull/202/commits/20fae5a64590785abc8512c76fda4f681bbab510)

File: GetMessage/handler.ts (hunk 3 above)
Comment (gquadrati) [reply GetMessage/handler.ts:151]: You're right, done in [20fae5a](https://github.com/pagopa/io-functions-services/pull/202/commits/20fae5a64590785abc8512c76fda4f681bbab510)

File: GetMessage/handler.ts (hunk 4 above)
Comment (gquadrati) [reply GetMessage/handler.ts]: Nice catch! Done in [20fae5a](https://github.com/pagopa/io-functions-services/pull/202/commits/20fae5a64590785abc8512c76fda4f681bbab510)

File: GetMessage/handler.ts (hunk 5 above)
Comment (gquadrati) [reply GetMessage/handler.ts]: Way better, thanks. Done in [a0a83cf](https://github.com/pagopa/io-functions-services/pull/202/commits/a0a83cf33f3d0c9340dbc1283dd95052a4e48f3b)

File: GetMessage/handler.ts [hunk 6]
```diff
@@ -252,19 +303,33 @@ export function GetMessageHandler(
...
+        b.fold(
+          () => messageWithoutAdvacedProperties,
+          () => ({
+            ...messageWithoutAdvacedProperties,
+            read_status: b.fold(
+              () => ReadStatusEnum.UNAVAILABLE,
+              () => getReadStatusForService(maybeMessageStatus)
+            )(canReadMessageReadStatus(auth.subscriptionId))
+          })
+        )(canReadAdvancedMessageInfo(message, auth.groups))
```
Comment (fabriziopapi) [inline GetMessage/handler.ts:334]:
```suggestion
        pipe(
          messageWithoutAdvacedProperties,
          E.fromPredicate(
            () => canReadAdvancedMessageInfo(message, auth.groups),
            () => messageWithoutAdvacedProperties
          ),
          E.map(() => canReadMessageReadStatus(auth.subscriptionId)),
          E.map(
            b.fold(
              () => ReadStatusEnum.UNAVAILABLE,
              () => getReadStatusForService(maybeMessageStatus)
            )
          ),
          E.map(read_status => ({
            ...messageWithoutAdvacedProperties,
            read_status
          })),
          E.toUnion
        )
```
we could flatten the nested fold using an either, but i'm not sure witch one is more readable

Comment (gquadrati) [reply GetMessage/handler.ts:334]: The one with less row, in my opinion :D

**Review by fabriziopapi — APPROVED** [review]

File: GetMessage/__tests__/handler.test.ts [hunk 7]
```diff
@@ -47,6 +47,9 @@ import { FeatureLevelTypeEnum } from "@pagopa/io-functions-commons/dist/generate
 
 import * as TE from "fp-ts/lib/TaskEither";
 
+import { ReadStatusEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/ReadStatus";
+import { PaymentStatusEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/PaymentStatus";
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply GetMessage/__tests__/handler.test.ts]: You're right, removed in [7965a98](https://github.com/pagopa/io-functions-services/pull/202/commits/7965a98012c42a6bdb2dc4156f7aa7317bedcdef) thanks!

**Pull request merged**